languages/python/sqlalchemy-oso/sqlalchemy_oso/hooks.py

```diff
@@ -0,0 +1,158 @@
+"""SQLAlchemy hooks that transparently enable oso on SQLAlchemy operations.
+
+There are several potential interfaces to integrate oso with SQLAlchemy:
+
+    - :py:func:`authorized_sessionmaker`: (**recommended**) Session factory that
+       creates a session that applies authorization on every query.
+    - :py:func:`enable_hooks`: Globally enable oso on all queries.
+    - :py:func:`make_authorized_query_cls`: Make an query class that is
+       authorized before execution.
+
+.. note::
+
+    If using any API besides :py:func:`authorized_sessionmaker`, ensure you set
+    ``enable_baked_queries=False`` on the session. Query caching can interfere
+    with authorization.
+
+    It is recommended to scope authorization context (the oso instance, user and
+    action) to a single session.  Otherwise, the identity map (SQLAlchemy's
+    cache of retrieved objects) may contain objects that were authorized for a
+    previous user. This could cause incorrect behavior.
+
+    :py:func:`authorized_sessionmaker` will enforce this.  If the authorization
+    context changes during the session, an Exception will be raised.
+"""
+import functools
+from typing import Any, Callable
+
+from sqlalchemy.event import listen, remove
+from sqlalchemy.orm.query import Query
+from sqlalchemy.orm import aliased, sessionmaker
+
+from oso import Oso
+
+from sqlalchemy_oso.auth import authorize_model, authorize_model_filter
+
+
+def enable_hooks(
+    get_oso: Callable[[], Oso],
+    get_user: Callable[[], Any],
+    get_action: Callable[[], Any],
+    target=None,
+):
+    """Enable all SQLAlchemy hooks."""
+    if target is None:
+        target = Query
+
+    disable_before_compile = enable_before_compile(
+        target, get_oso, get_user, get_action
+    )
+
+    def disable_hooks():
+        disable_before_compile()
+
+    return disable_hooks
+
+
+def enable_before_compile(
+    target,
+    get_oso: Callable[[], Oso],
+    get_user: Callable[[], Any],
+    get_action: Callable[[], Any],
+):
+    """Enable before compile hook."""
+    auth = functools.partial(
+        authorize_query, get_oso=get_oso, get_user=get_user, get_action=get_action
+    )
+
+    listen(target, "before_compile", auth, retval=True)
+
+    return lambda: remove(target, "before_compile", auth)
+
+
+def authorize_query(query: Query, get_oso, get_user, get_action) -> Query:
+    """Authorize an existing query with an oso instance, user and action."""
+    oso = get_oso()
+    action = get_action()
+    actor = get_user()
+
+    # TODO (dhatch): Super hack to get this to work in flask... fix this.
+    query = query.enable_assertions(False)
+
+    entities = {column["entity"] for column in query.column_descriptions}
```

Comment:
Add `if column["entity"]` guard into comprehension instead of the `if entity is None:` check below?

Reply:
I think it's more readable as is with the comment.